obj_factory.h

```diff
@@ -0,0 +1,129 @@
+/* Copyright © 2001-2015, Canal TP and/or its affiliates. All rights reserved.
+
+This file is part of Navitia,
+    the software to build cool stuff with public transport.
+
+Hope you'll enjoy and contribute to this project,
+    powered by Canal TP (www.canaltp.fr).
+Help us simplify mobility and open public transport:
+    a non ending quest to the responsive locomotion way of traveling!
+
+LICENCE: This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+Stay tuned using
+twitter @navitia
+IRC #navitia on freenode
+https://groups.google.com/d/forum/navitia
+www.navitia.io
+*/
+#pragma once
+#include <vector>
+#include <unordered_map>
+#include "utils/functions.h"
+#include "utils/idx_map.h"
+
+namespace navitia{
+
+/*
+ * Factory handling all collections for one object
+ * Object must inherit from navitia::type::Header
+ */
+template<typename ObjType>
+class ObjFactory {
+public:
+    typedef typename std::vector<std::unique_ptr<ObjType>>::iterator iterator;
+    typedef typename std::vector<std::unique_ptr<ObjType>>::const_iterator const_iterator;
+
+    template<typename ...Args>
+    ObjType* emplace(const std::string& uri, Args&& ...args) {
+        check_free_uri(uri);
+        vec.push_back(std::make_unique<ObjType>(std::forward<Args>(args)...));
```

Comment:
why do you use push_back and not emplace_back ?

the mechanism does not seem  straight forward.
wouldn't it be better for you to create the unique_ptr and pass it to a `save` method (your `manage_vec_back_consistency`) ?